ethpm/manifest.py

```diff
@@ -0,0 +1,257 @@
+        - `links`           Dict[str,str]
+            Recommended keys for `links`
+            - `website`
+            - `documentation`
+            - `repository`
+
+    :compiled_json_output_path:         [optional]
+        string of the path to standard-json compiler output
+        compiler output *must* be generated with `abi` & `devdoc` flags enabled
+
+            i.e.
+            `solc --output-dir ./ --combined-json
+            abi,bin,bin-runtime,devdoc,interface,userdoc contracts/Owned.sol`
+
+    :contract_types:    [optional]
+        List[str] containing the contract types to be included in this manifest.
+            * Packages should only include contract types that can be
+              found in the source files for this package.
+            * Packages should not include abstract contracts in the contract types section.
+            * Packages should not include contract types from dependencies.
+
+    :build_dependencies:        [optional]
+        Dict[str, str] with the package name pointing to a content-
+        addressable URI which resolves to the package's manifest.
+
+    :allow_extra_meta_fields:   [optional]
+        Flag that must be explicitly set to true if user wants to allow
+        fields in the meta object that are not listed in the spec.
+    """
+    validate_manifest_version(manifest_version)
+    validate_package_name(package_name)
+    validate_package_version(version)
+    manifest = {
+        "manifest_version": manifest_version,
+        "package_name": package_name,
+        "version": version,
+    }
+    build_fns = (
+        build_meta(meta, allow_extra_meta_fields),
+        build_sources(compiled_json_output_path),
+        build_contract_types(compiled_json_output_path, contract_types),
+        build_build_dependencies(build_dependencies),
+    )
+    built_manifest = cytoolz.pipe(manifest, *build_fns)
+    return built_manifest
+
+
+#
+# Meta
+#
+
+
+@cytoolz.curry
+def build_meta(
+    meta: Dict[str, Any], allow_extra_meta_fields: bool, manifest: Dict[str, Any]
+) -> Dict[str, Any]:
+    """
+    Validate meta object keys and values, and attach object to manifest if valid.
+    """
+
+    if meta:
+        validate_meta_object(meta, allow_extra_meta_fields)
+        new_manifest = cytoolz.assoc(manifest, "meta", meta)
+        return new_manifest
+    return manifest
+
+
+def validate_meta_object(meta: Dict[str, Any], allow_extra_meta_fields: bool) -> None:
+    """
+    Validates that every key is one of `META_FIELDS` and has a value of the expected type.
+    """
+    for key, value in meta.items():
+        if key in META_FIELDS:
+            if type(value) is not META_FIELDS[key]:
+                raise ValidationError(
+                    "Values for {0} are expected to have the type {1}, instead got {2}.".format(
+                        key, META_FIELDS[key], type(value)
+                    )
+                )
+            continue
+        if allow_extra_meta_fields:
```

Comment:
This should be an `elif`.  The current code is functionally equivalent since the previous `if` block has a `continue` but a refactor could change that and cause breakage.